### PR TITLE
2.40 dhis2 14539

### PR DESCRIPTION
**Pull request by d-bernat**
please see: https://dhis2.atlassian.net/browse/DHIS2-14539

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 1]
```diff
@@ -1237,6 +1262,32 @@ private String nullAndEmptyMatcher( QueryItem item, QueryFilter filter, String f
         }
     }
 
+    /**
+     * The tool for merging of query items relevant for order by DML command
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java:1266]:
```suggestion
     * Method responsible for merging query items based on sorting parameters.
```

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 2]
```diff
@@ -417,12 +418,36 @@ else if ( queryItem.getValueType() == ValueType.NUMBER && !isGroupByClause )
...
+        {
+            return getColumnAndAliasWithNullIfFunction( queryItem );
+        }
         else
         {
             return getColumnAndAlias( queryItem, isGroupByClause, "" );
         }
     }
 
+    private ColumnAndAlias getColumnAndAliasWithNullIfFunction( QueryItem queryItem )
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java:431]: Would be nice to document(Javadoc) this method and also give an example of the output expected.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/ColumnWithNullIfAndAlias.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.event.data;
+
+import org.apache.commons.lang3.StringUtils;
+
+final class ColumnWithNullIfAndAlias extends ColumnAndAlias
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/ColumnWithNullIfAndAlias.java:32]: Please add a Javadoc for this class and its methods.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 4]
```diff
@@ -417,12 +418,36 @@ else if ( queryItem.getValueType() == ValueType.NUMBER && !isGroupByClause )
...
+        if ( queryItem.hasProgramStage() && queryItem.getItem().getDimensionItemType() == DATA_ELEMENT )
+        {
+            return ColumnWithNullIfAndAlias.ofColumnWithNullIfAndAlias( column,
+                queryItem.getProgramStage().getUid() + "." + queryItem.getItem().getUid() );
+        }
+
+        return ColumnWithNullIfAndAlias.ofColumnWithNullIfAndAlias( column, queryItem.getItem().getUid() );
+    }
+
+    private boolean hasOrderByClauseQueryItem( QueryItem queryItem, EventQueryParams params )
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java:444]:
```suggestion
    private boolean hasOrderByClauseForQueryItem( QueryItem queryItem, EventQueryParams params )
```

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/ColumnWithNullIfAndAlias.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+{
+    private ColumnWithNullIfAndAlias( String column, String alias )
+    {
+        super( column, alias );
+    }
+
+    static ColumnWithNullIfAndAlias ofColumnWithNullIfAndAlias( String column, String alias )
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/ColumnWithNullIfAndAlias.java:39]: Please unit test this method.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/ColumnWithNullIfAndAlias.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        return new ColumnWithNullIfAndAlias( column, alias );
+    }
+
+    @Override
+    public String asSql()
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/ColumnWithNullIfAndAlias.java:45]: Please unit test this method.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 7]
```diff
@@ -1237,6 +1262,32 @@ private String nullAndEmptyMatcher( QueryItem item, QueryFilter filter, String f
...
     }
 
+    /**
+     * The tool for merging of query items relevant for order by DML command
+     *
+     * @param params the {@link EventQueryParams} to drive the query item list
+     *        generation.
+     * @return the distinct {@link List<QueryItem>} relevant for order by DML.
+     */
+    private static List<QueryItem> getDistinctOrderByColumns( EventQueryParams params )
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java:1272]: Does it need to be static?

**Review by maikelarabori — COMMENTED** [review]
Thanks @d-bernat, please review my comments.

Cheers

**Review by maikelarabori — APPROVED** [review]

**Review by gnespolino — APPROVED** [review]

**Pull request merged**